Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,17 +14,56 @@
 
 import Crdkafka
 import Logging
+import NIOConcurrencyHelpers
 import NIOCore
 
+/// `NIOAsyncSequenceProducerBackPressureStrategy` that always returns true.
+struct NoBackPressure: NIOAsyncSequenceProducerBackPressureStrategy {
+    func didYield(bufferDepth: Int) -> Bool { true }
+    func didConsume(bufferDepth: Int) -> Bool { true }
+}
+
+// MARK: - ShutDownOnTerminate
+
+/// `NIOAsyncSequenceProducerDelegate` that terminates the shuts the producer down when
+/// `didTerminate()` is invoked.
+struct ShutdownOnTerminate: @unchecked Sendable { // We can do that because our stored propery is protected by a lock
+    let stateMachine: NIOLockedValueBox<KafkaProducer.StateMachine>
+}
+
+extension ShutdownOnTerminate: NIOAsyncSequenceProducerDelegate {
+    func produceMore() {
+        // No back pressure
+        return
+    }
+
+    func didTerminate() {
+        let action = self.stateMachine.withLockedValue { $0.finish() }
+        switch action {
+        case .shutdownGracefullyAndFinishSource(let client, let source, let topicHandles):
+            Task {
```

Comment:
We can't do this. Can we instead do the same thing for flush as you did with commitSync and close i.e. using the queue and a callback. Then we can use a continuation somewhere. We somehow need to get this termination event over to the `KafkaProducer`s run method probably.

Reply:
Unfortunately, I haven't found any async alternative to [`rd_kafka_flush`](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#aaff06c4372bce917c17f3c1a5d8b205d) like there is with [`rd_kafka_consumer_close_queue`](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#a9dd5c18bdfed81c8847b259f0a8d498d) or [`rd_kafka_commit_queue`](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#af76a6a73baa9c2621536e3f6882a3c1a).

What I have done however is get rid of [`rd_kafka_flush`](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#aaff06c4372bce917c17f3c1a5d8b205d) and imitate this method as good as possible using our `StateMachine` and our run loop. However, what I have done is very basic compared to the [actual `rd_kafka_flush` implementation](https://github.com/confluentinc/librdkafka/blob/d174c0dddc1a6c17a93b9b3e1d10e5cf4b2b61b2/src/rdkafka.c#L4346). I am basically just polling until there are no more events in `librdkafka`'s outgoing queue when `KafkaProducer.shutdownGracefully` has been invoked. Atm there are no `timeouts` whatsoever. Please let me know what you think.